libbeat/publisher/queue/diskqueue/compression_test.go

```diff
@@ -0,0 +1,157 @@
+		},
+		"abc pierrec lz4": {
+			plaintext: []byte("abc"),
+			compressed: []byte{
+				0x04, 0x22, 0x4d, 0x18,
+				0x64, 0x70, 0xb9, 0x03,
+				0x00, 0x00, 0x80, 0x61,
+				0x62, 0x63, 0x00, 0x00,
+				0x00, 0x00, 0xff, 0x53,
+				0xd1, 0x32},
+		},
+	}
+
+	for name, tc := range tests {
+		dst := make([]byte, len(tc.plaintext))
+		src := bytes.NewReader(tc.compressed)
+		cr := NewCompressionReader(io.NopCloser(src))
+		n, err := cr.Read(dst)
+		assert.Nil(t, err, name)
+		assert.Equal(t, len(tc.plaintext), n, name)
+		assert.Equal(t, tc.plaintext, dst, name)
+	}
+}
+
+func TestCompressionWriter(t *testing.T) {
+	tests := map[string]struct {
+		plaintext  []byte
+		compressed []byte
+	}{
+		"abc pierrec lz4": {
+			plaintext: []byte("abc"),
+			compressed: []byte{
+				0x04, 0x22, 0x4d, 0x18,
+				0x64, 0x70, 0xb9, 0x03,
+				0x00, 0x00, 0x80, 0x61,
+				0x62, 0x63, 0x00, 0x00,
+				0x00, 0x00, 0xff, 0x53,
+				0xd1, 0x32},
+		},
+	}
+
+	for name, tc := range tests {
+		var dst bytes.Buffer
+		cw := NewCompressionWriter(NopWriteCloseSyncer(NopWriteCloser(&dst)))
+		n, err := cw.Write(tc.plaintext)
+		cw.Close()
+		assert.Nil(t, err, name)
+		assert.Equal(t, len(tc.plaintext), n, name)
+		assert.Equal(t, tc.compressed, dst.Bytes(), name)
+	}
+}
+
+func TestCompressionRoundTrip(t *testing.T) {
+	tests := map[string]struct {
+		plaintext []byte
+	}{
+		"no repeat":  {plaintext: []byte("abcdefghijklmnopqrstuvwxzy01234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ")},
+		"256 repeat": {plaintext: []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")},
+	}
+	for name, tc := range tests {
+		pr, pw := io.Pipe()
+		src := bytes.NewReader(tc.plaintext)
+		var dst bytes.Buffer
+
+		go func() {
+			cw := NewCompressionWriter(NopWriteCloseSyncer(pw))
+			_, err := io.Copy(cw, src)
+			assert.Nil(t, err, name)
+			cw.Close()
+		}()
+
+		cr := NewCompressionReader(pr)
+		_, err := io.Copy(&dst, cr)
+		assert.Nil(t, err, name)
+		assert.Equal(t, tc.plaintext, dst.Bytes(), name)
+	}
+}
+
+func TestCompressionSync(t *testing.T) {
+	tests := map[string]struct {
+		plaintext []byte
+	}{
+		"no repeat":  {plaintext: []byte("abcdefghijklmnopqrstuvwxzy01234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ")},
+		"256 repeat": {plaintext: []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")},
+	}
+	for name, tc := range tests {
+		pr, pw := io.Pipe()
+		var dst bytes.Buffer
+		go func() {
+			cw := NewCompressionWriter(NopWriteCloseSyncer(pw))
+			src1 := bytes.NewReader(tc.plaintext)
+			_, err := io.Copy(cw, src1)
+			assert.Nil(t, err, name)
+			err = cw.Sync()
```

Comment:
What is the sync call doing here? Finalizing a block? 

Could you add a comment explaining what this test is for? I remember there was an issue with the sync call this is probably testing, but a comment on this would help future developers be aware of any potential issues here.

Reply:
good idea, added comment